spec.bs

```diff
@@ -501,14 +506,57 @@ are:
     1. Let |permission| be the result of [=checking interest group permissions=] with
       |owner|, |frameOrigin|, and "`leave`".
     1. If |permission| is false, then [=queue a task=] to [=reject=] |p| with a
-       "{{NotAllowedError}}" {{DOMException}} and do not run the remaining steps.
+       "{{NotAllowedError}}" {{DOMException}} and abort these steps.
     1. [=Queue a task=] to [=resolve=] |p| with `undefined`.
     1. [=list/Remove=] [=interest groups=] from the user agent's [=interest group set=] whose
       [=interest group/owner=] is |owner| and [=interest group/name=] is |name|.
 1. Return |p|.
 
 </div>
 
+<h3 id="clearoriginjoinedAdInterestGroups">clearOriginJoinedAdInterestGroups()</h3>
+{{Window/navigator}}.{{Navigator/clearOriginJoinedAdInterestGroups()}} removes a user from
+[=interest groups=] whose [=interest group/joining origin=] is the current top level page's
+origin.
+
+
+<xmp class="idl">
+[SecureContext]
+partial interface Navigator {
+  Promise<undefined> clearOriginJoinedAdInterestGroups(
+      USVString owner, optional sequence<USVString> interestGroupsToKeep = []);
+};
+</xmp>
+
+<div algorithm>
+
+The <dfn for=Navigator method>clearOriginJoinedAdInterestGroups(|owner|, |interestGroupsToKeep|)</dfn>
+method steps are:
+
+1. Let |frameOrigin| be [=this=]'s [=relevant settings object=]'s
+  [=environment settings object/origin=].
+1. [=Assert=] that |frameOrigin| is not an [=opaque origin=] and its [=origin/scheme=] is "`https`".
+1. Let |p| be [=a new promise=].
+1. If [=this=]'s [=relevant global object=]'s [=associated Document=] is not [=allowed to use=] the
+  "[=join-ad-interest-group=]" [=policy-controlled feature=], then [=exception/throw=] a
+  "{{NotAllowedError}}" {{DOMException}}.
+
+  Note: Both joining and leaving interest groups use the "join-ad-interest-group" feature.
+1. Let |ownerOrigin| be the result of [=parsing an https origin=] with |owner|.
+1. If |ownerOrigin| is failure, [=exception/throw=] a {{TypeError}}.
+1. Run these steps [=in parallel=]:
+  1. Let |permission| be the result of [=checking interest group permissions=] with
+    |ownerOrigin|, |frameOrigin|, and "`leave`".
+  1. If |permission| is false, then [=queue a task=] to [=reject=] |p| with a

```

Comment:
See https://github.com/WICG/turtledove/pull/796#discussion_r1365604036 and the resulting fixed code which this should model: https://github.com/WICG/turtledove/pull/796/files#diff-6f5a1d8263b0b0c42e2716ba5750e3652e359532647ac934c1c70086ae3ceddaR2200-R2201.

Reply:
That sounds like it would not be accurate to me.  Task queues are associated with an event loop, and event loops are destroyed when all associated browsing contexts are destroyed.  These tasks not cancelled when the frame is destroyed (e.g., if it's destroyed while we're fetching the .well-known file).  We continue fetching the file, and clear the interest groups according to the result, even if the frame has gone away.  We're also not run on a serialize event loop associated with the browsing context, so even without that, I'm not sure it would be terribly accurate.

Reply:
I’m not following. I’m not saying to cancel any tasks or fetching. I’m just saying that Promises are tied to an event loop, and as such they can only be canceled or fulfilled on that same event loop.

Reply:
Sorry, I completely misunderstood.  I thought the task was for the events done in parallel, rather than for running the promise.  I've address the comment.  Let me know if I did anything wrong, thanks!

Reply:
And to add - your comment was pretty clear, and made on the exact right line.  I'm not sure how I managed to misinterpret it.